From 9ed6ee4044ca8a49f17c50c8c7a5225d40e415c5 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Wed, 19 Nov 2025 15:56:17 +0100
Subject: [PATCH] usb: dwc2: add an optional phy_reset when starting

On STM32MP21, the platform may go to stop mode at runtime, using
psci. During stop mode, the PHY clock source will turn off. In
case the dwc2 wasn't suspending before, the PHY needs to be reset,
when starting the gadget (udc_start), or in the drd layer, upon
usb role switch routine. Add an optional parameter to handle it.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Ifad54bb16f2199a63e2c652e5bb7cd32e3ff7d37
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512985
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/dwc2/core.h   |  1 +
 drivers/usb/dwc2/drd.c    | 12 ++++++++++++
 drivers/usb/dwc2/gadget.c | 11 +++++++++++
 drivers/usb/dwc2/params.c |  1 +
 4 files changed, 25 insertions(+)

--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -509,6 +509,7 @@ struct dwc2_core_params {
 	u32 g_tx_fifo_size[MAX_EPS_CHANNELS];
 
 	bool change_speed_quirk;
+	bool reset_phy_on_start;
 };
 
 /**
--- a/drivers/usb/dwc2/drd.c
+++ b/drivers/usb/dwc2/drd.c
@@ -200,6 +200,18 @@ static int dwc2_drd_role_sw_set(struct u
 		if (hsotg->params.power_down == DWC2_POWER_DOWN_PARAM_NONE &&
 		    !hsotg->params.no_clock_gating)
 			dwc2_gadget_exit_clock_gating(hsotg, 0);
+
+		if (hsotg->params.reset_phy_on_start) {
+			spin_unlock_irqrestore(&hsotg->lock, flags);
+			/*
+			 * In case the gadget has been stopped earlier, and
+			 * the platform enters some low power states, the phy
+			 * may need a reset before starting again.
+			 */
+			if (phy_reset(hsotg->phy))
+				dev_warn(hsotg->dev, "PHY reset failed\n");
+			spin_lock_irqsave(&hsotg->lock, flags);
+		}
 	}
 
 	if (role == USB_ROLE_HOST) {
--- a/drivers/usb/dwc2/gadget.c
+++ b/drivers/usb/dwc2/gadget.c
@@ -4685,6 +4685,17 @@ static int dwc2_hsotg_udc_start(struct u
 			goto err;
 	}
 
+	if (hsotg->params.reset_phy_on_start) {
+		/*
+		 * In case the gadget has been stopped earlier, and
+		 * the platform enters some low power states, the phy
+		 * may need a reset before starting again.
+		 */
+		ret = phy_reset(hsotg->phy);
+		if (ret)
+			dev_warn(hsotg->dev, "PHY reset failed\n");
+	}
+
 	if (!IS_ERR_OR_NULL(hsotg->uphy))
 		otg_set_peripheral(hsotg->uphy->otg, &hsotg->gadget);
 
--- a/drivers/usb/dwc2/params.c
+++ b/drivers/usb/dwc2/params.c
@@ -285,6 +285,7 @@ static void dwc2_set_stm32mp21_hsotg_par
 	p->activate_stm32_bvaloval_en = true;
 	p->activate_stm32_otgarcr_en = true;
 	p->stm32_has_batt_chg_det = true;
+	p->reset_phy_on_start = true;
 }
 
 const struct of_device_id dwc2_of_match_table[] = {
